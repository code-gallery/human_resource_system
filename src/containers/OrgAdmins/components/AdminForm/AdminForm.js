import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Notifications, { notify } from 'react-notify-toast'
import Select from 'react-select'
import { NOTIFICATION_TIMEOUT } from 'containers/constants'

class AdminForm extends Component {
  constructor(props) {
    super(props)
    this.state = {
      selectedAdmin: {
        user: null,
        primary: null
      }
    }
  }

  notifyFn(message, type) {
    // this is a wrapper function to make the unit test happy
    notify.show(message, type, NOTIFICATION_TIMEOUT)
  }

  optionRenderer = (option) => {
    return (
      <div className="institution-select">
        <img src={option.profile_image} alt={option.first_name} />
        <p>{`${option.first_name} ${option.last_name}`}</p>
      </div>
    )
  }

  selectAdmin = (value) => {
    this.setState({
      selectedAdmin: {
        user: value
      }
    })
  }

  isSelectedAdminPrimary = (value) => {
    this.setState({
      selectedAdmin: {
        ...this.state.selectedAdmin,
        primary: value
      }
    })
  }

  addAdminHandler = (e) => {
    e.preventDefault()

    const { selectedAdmin } = this.state

    if (!selectedAdmin.user) {
      this.notifyFn('Select user first', 'error')
    } else {
      this.props.addAdmin({
        orgId: this.props.orgId,
        user_id: selectedAdmin.user.id,
        primary: selectedAdmin.primary && selectedAdmin.primary.label === 'Yes'
      })
    }
  }

  render() {
    const { showForm, employees, closeForm } = this.props
    const { user, primary } = this.state.selectedAdmin

    if (!showForm) {
      return null
    }

    return (
      <form className="admin">
        <Notifications />

        <div className="col-md-3">
          <div className="form-group">
            <label>Users</label>
            <Select
              options={employees}
              labelKey="first_name"
              valueKey="first_name"
              value={{ first_name: user ? `${user.first_name} ${user.last_name}` : 'Select...' }}
              optionRenderer={this.optionRenderer}
              onChange={this.selectAdmin}
            />
          </div>
        </div>

        <div className="col-md-3">
          <div className="form-group">
            <label>Primary Contact</label>
            <Select
              valueKey="label"
              value={{ label: primary ? primary.label : 'Select...' }}
              options={[ { label: 'Yes', value: 'Yes' }, { label: 'No', value: 'No' } ]}
              onChange={this.isSelectedAdminPrimary}
              disabled={!user} />
          </div>
        </div>

        <div className="col-md-6">
          <button className="btn green-btn admin-form-btn" onClick={this.addAdminHandler}>
            Confirm
          </button>

          <button className="btn grey-btn admin-form-btn admin-form-btn" onClick={closeForm}>
            Cancel
          </button>
        </div>
      </form>
    )
  }
}

AdminForm.propTypes = {
  showForm: PropTypes.bool,
  orgId: PropTypes.string,
  employees: PropTypes.array,
  closeForm: PropTypes.func.isRequired,
  addAdmin: PropTypes.func.isRequired
}

export default AdminForm
