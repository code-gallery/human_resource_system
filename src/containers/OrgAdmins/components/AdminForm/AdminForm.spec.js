import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import AdminForm from './AdminForm'

describe('containers/OrgAdmins/components/AdminForm/AdminForm', () => {
  let fullProps
  let commonProps

  beforeEach(() => {
    commonProps = {
      closeForm: jest.fn(),
      addAdmin: jest.fn()
    }
    fullProps = {
      ...commonProps,
      showForm: true,
      orgId: '142',
      employees: []
    }
  })

  describe('when props.showForm is false', () => {
    it('renders null', () => {
      const tree = renderer.create(
        <AdminForm {...commonProps} />
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })
  })

  describe('when props.showForm is true', () => {
    it('renders correctly', () => {
      const tree = renderer.create(
        <AdminForm {...fullProps} />
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })
  })

  describe('<AdminForm', () => {
    let wrapper
    beforeEach(() => {
      wrapper = shallow(
        <AdminForm {...fullProps} />
      )
      jest.spyOn(wrapper.instance(), 'notifyFn').mockImplementation(function() {})
    })

    describe('optionRenderer', () => {
      it('renders correct element', () => {
        const option = {
          profile_image: 'profile_image',
          first_name: 'Jane',
          last_name: 'Doe'
        }

        const tree = renderer.create(
          wrapper.instance().optionRenderer(option)
        ).toJSON()

        expect(tree).toMatchSnapshot()
      })
    })

    describe('selectAdmin', () => {
      describe('when value is valid', () => {
        it('returns correctly', () => {
          const user = 'user1'
          wrapper.instance().selectAdmin(user)
          expect(wrapper.state()).toEqual(expect.objectContaining({
            selectedAdmin: {
              user: user
            }
          }))
        })
      })
    })

    describe('isSelectedAdminPrimary', () => {
      describe('when value is valid', () => {
        it('returns correctly', () => {
          const value = 'primary1'
          wrapper.instance().isSelectedAdminPrimary(value)
          expect(wrapper.state().selectedAdmin).toEqual(expect.objectContaining({
            primary: value
          }))
        })
      })
    })

    describe('addAdminHandler', () => {
      describe('when state.selectedAdmin.user is not valid', () => {
        it('shows an error notification', () => {
          wrapper.instance().addAdminHandler({ preventDefault: jest.fn() })
          expect(wrapper.instance().notifyFn)
            .toHaveBeenCalledWith(expect.any(String), 'error')
        })
      })

      describe('when state.selectedAdmin.user is valid', () => {
        it('calls props.addAdmin correctly (1)', () => {
          const state = {
            selectedAdmin:
              {
                user: { id: 'userid1' },
                primary: { label: 'Yes' }
              }
          }
          wrapper.setState(state)
          wrapper.instance().addAdminHandler({ preventDefault: jest.fn() })
          expect(fullProps.addAdmin)
            .toHaveBeenCalledWith({
              orgId: fullProps.orgId,
              user_id: state.selectedAdmin.user.id,
              primary: true
            })
        })

        it('calls props.addAdmin correctly (2)', () => {
          const state = {
            selectedAdmin:
              {
                user: { id: 'userid1' },
                primary: { label: 'No' }
              }
          }
          wrapper.setState(state)
          wrapper.instance().addAdminHandler({ preventDefault: jest.fn() })
          expect(fullProps.addAdmin)
            .toHaveBeenCalledWith({
              orgId: fullProps.orgId,
              user_id: state.selectedAdmin.user.id,
              primary: false
            })
        })
      })
    })
  })
})
