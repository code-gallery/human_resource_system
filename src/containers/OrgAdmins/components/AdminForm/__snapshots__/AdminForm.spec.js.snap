// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/OrgAdmins/components/AdminForm/AdminForm <AdminForm optionRenderer renders correct element 1`] = `
<div
  className="institution-select"
>
  <img
    alt="Jane"
    src="profile_image"
  />
  <p>
    Jane Doe
  </p>
</div>
`;

exports[`containers/OrgAdmins/components/AdminForm/AdminForm when props.showForm is false renders null 1`] = `null`;

exports[`containers/OrgAdmins/components/AdminForm/AdminForm when props.showForm is true renders correctly 1`] = `
<form
  className="admin"
>
  <div
    id="notification-wrapper"
  />
  <div
    className="col-md-3"
  >
    <div
      className="form-group"
    >
      <label>
        Users
      </label>
      <div
        className="Select has-value is-clearable is-searchable Select--single"
        style={undefined}
      >
        <div
          className="Select-control"
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          style={undefined}
        >
          <span
            className="Select-multi-value-wrapper"
            id="react-select-2--value"
          >
            <div
              className="Select-value"
              style={undefined}
              title={undefined}
            >
              <span
                aria-selected="true"
                className="Select-value-label"
                id="react-select-2--value-item"
                role="option"
              >
                Select...
              </span>
            </div>
            <div
              className="Select-input"
              style={
                Object {
                  "display": "inline-block",
                }
              }
            >
              <input
                aria-activedescendant="react-select-2--value"
                aria-describedby={undefined}
                aria-expanded="false"
                aria-haspopup="false"
                aria-label={undefined}
                aria-labelledby={undefined}
                aria-owns=""
                className={undefined}
                id={undefined}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                role="combobox"
                style={
                  Object {
                    "boxSizing": "content-box",
                    "width": "5px",
                  }
                }
                tabIndex={undefined}
                value=""
              />
              <div
                style={
                  Object {
                    "height": 0,
                    "left": 0,
                    "overflow": "scroll",
                    "position": "absolute",
                    "top": 0,
                    "visibility": "hidden",
                    "whiteSpace": "pre",
                  }
                }
              >
                
              </div>
            </div>
          </span>
          <span
            aria-label="Clear value"
            className="Select-clear-zone"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            title="Clear value"
          >
            <span
              className="Select-clear"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&times;",
                }
              }
            />
          </span>
          <span
            className="Select-arrow-zone"
            onMouseDown={[Function]}
          >
            <span
              className="Select-arrow"
              onMouseDown={[Function]}
            />
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="col-md-3"
  >
    <div
      className="form-group"
    >
      <label>
        Primary Contact
      </label>
      <div
        className="Select has-value is-clearable is-disabled is-searchable Select--single"
        style={undefined}
      >
        <div
          className="Select-control"
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          style={undefined}
        >
          <span
            className="Select-multi-value-wrapper"
            id="react-select-3--value"
          >
            <div
              className="Select-value"
              style={undefined}
              title={undefined}
            >
              <span
                aria-selected="true"
                className="Select-value-label"
                id="react-select-3--value-item"
                role="option"
              >
                Select...
              </span>
            </div>
            <div
              aria-activedescendant="react-select-3--value"
              aria-disabled="true"
              aria-expanded={false}
              aria-label={undefined}
              aria-labelledby={undefined}
              aria-owns=""
              className="Select-input"
              onBlur={[Function]}
              onFocus={[Function]}
              role="combobox"
              style={
                Object {
                  "border": 0,
                  "display": "inline-block",
                  "width": 1,
                }
              }
              tabIndex={0}
            />
          </span>
          <span
            className="Select-arrow-zone"
            onMouseDown={[Function]}
          >
            <span
              className="Select-arrow"
              onMouseDown={[Function]}
            />
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="col-md-6"
  >
    <button
      className="btn green-btn admin-form-btn"
      onClick={[Function]}
    >
      Confirm
    </button>
    <button
      className="btn grey-btn admin-form-btn admin-form-btn"
      onClick={[Function]}
    >
      Cancel
    </button>
  </div>
</form>
`;
